### PR TITLE
tcp_boundary: drop source correctly

### PR DESCRIPTION
**Pull request by antiguru**
Instead of waiting to schedule ok and err streams to drop the source, only
require a single of the streams to be terminate to terminate both. This
prevents a situation where the ok stream is terminated, but the err stream
will not pick up the termination because it won't be scheduled anymore.

Signed-off-by: Moritz Hoffmann <mh@materialize.com>

### Motivation
  * This PR fixes a previously unreported bug: Sources aren't cleaned up after compute.

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by antiguru** [comment]
This is an alternative to #11945.

File: src/storage/src/boundary/tcp_boundary.rs [hunk 1]
```diff
@@ -336,28 +338,29 @@ pub mod server {
...
                 convert,
-                token: Some(token),
+                token,
             }
         }
     }
 
     impl<T, D, R, F: Fn(EventCore<T, D>) -> R> EventPusherCore<T, D> for UnboundedEventPusher<R, F> {
         fn push(&mut self, event: EventCore<T, D>) {
+            let mut token = self.token.borrow_mut();
```
Comment (guswynn) [inline src/storage/src/boundary/tcp_boundary.rs:356]: is there any chance this panics? can both pushers be called at the same time?

**Review by guswynn — COMMENTED** [review]
I am convinced now that https://github.com/MaterializeInc/materialize/pull/11945 is definitely a backwards approach, but this technique concerns me for 2 reasons:

- What is different about the `ok` and `err` stream where, as the source is dying? Is the `ok` stream always going to get a push? if it is also not getting any data (just like the `err`) stream, could neither of these be called and we have the same bug?
- My other comment inline; I don't understand timely well enough to know if this is a concern

File: src/storage/src/boundary/tcp_boundary.rs (hunk 1 above)
Comment (antiguru) [reply src/storage/src/boundary/tcp_boundary.rs:356]: Nope, they run on the same thread!

**Comment by guswynn** [comment]
https://github.com/MaterializeInc/materialize/pull/12082 should be able to replace this pr

**Comment by frankmcsherry** [comment]
On my list of "pending reviews" but I suspect we just want to close?

**Comment by guswynn** [comment]
@frankmcsherry yeah, this should be relevant anymore!

**Pull request closed without merging**